oqswrap.py

```diff
@@ -0,0 +1,327 @@
+# * add unit test
+
+# import ctypes to call native
+from ctypes import *
+# platform.system to learn the OS
+from platform import system
+
+# expected return value from native OQS functions
+_OQS_SUCCESS = 0
+
+# load native OQS library
+if system() == 'Windows':
+        liboqs = windll.LoadLibrary('oqs')
+else:
+    try:
+        # try to load a local library first
+        liboqs = cdll.LoadLibrary('./liboqs.so')
+    except OSError:
+        # no local liboqs, try to load the system one
+        liboqs = cdll.LoadLibrary('liboqs.so')
+
+class MechanismNotSupportedError(Exception):
+    """Exception raised when an algorithm is not supported by OQS.
+
+    Attribute:
+        alg_name -- the requested algorithm name
+    """
+
+    def __init__(self, alg_name):
+        self.alg_name = alg_name
+        self.message = alg_name + ' is not supported by OQS'
+    
+class MechanismNotEnabledError(MechanismNotSupportedError):
+    """Exception raised when an algorithm is not supported but not enabled by OQS.
+
+    Attribute:
+        alg_name -- the requested algorithm name
+    """
+
+    def __init__(self, alg_name):
+        self.alg_name = alg_name
+        self.message = alg_name + ' is not supported but not enabled by OQS'
+
+############################################
+# KEM
+############################################
+
+# The native KEM structure returned by OQS
+class OQS_KEM(Structure):
+    _fields_ = [
+        ("method_name", c_char_p),
+        ("alg_version", c_char_p),
+        ("claimed_nist_level", c_ubyte),
+        ("ind_cca", c_ubyte),
+        ("length_public_key", c_size_t),
+        ("length_secret_key", c_size_t),
+        ("length_ciphertext", c_size_t),
+        ("length_shared_secret", c_size_t),
+        ("keypair_cb", c_void_p),
+        ("encaps_cb", c_void_p),
+        ("decaps_cb", c_void_p)
+    ]
+liboqs.OQS_KEM_new.restype = POINTER(OQS_KEM)
+
+liboqs.OQS_KEM_alg_identifier.restype = c_char_p
+
+# populate the list of enabled KEMs
+_max_number_KEMs = liboqs.OQS_KEM_alg_count()
+_supported_KEMs = []
+_enabled_KEMs = []
+for i in range(0, _max_number_KEMs):
+    algid =  liboqs.OQS_KEM_alg_identifier(i)
+    algid_string = algid.decode()
+    _supported_KEMs.append(algid_string)
+    # check if alg is enabled
+    is_supported = True
+    try:
+        kem = liboqs.OQS_KEM_new( create_string_buffer(algid) )
+        kem.contents # triggers ValueError if init failed # TODO: must be a better way
+        liboqs.OQS_KEM_free(kem)
+    except ValueError:
+        is_supported = False
+    if (is_supported):
+        _enabled_KEMs.append(algid_string)
+    
+def get_enabled_KEM_mechanisms():
+    """Returns the list of enabled KEM mechanisms."""
+    return _enabled_KEMs
+
+def print_enabled_KEM_mechanisms():
+    """Prints the list of enabled KEM mechanisms."""
+    print('Enabled KEM mechanisms:', ', '.join(_enabled_KEMs))
+
+class encap_data:
```

Comment:
I wasn't sure how to best return multiple values; I've seen various options proposed. I've considered a NamedTuple but wanted to avoid importing more packages. I'm happy to switch if it's a better Python style

Reply:
Given there's only 2 variables, the simplest is to delete the class & return a tuple for both return cases:
```
def encap_secret():
    # do stuff
    return bytes(ciphertext), bytes(shared_secret) if _rv == _OQS_SUCCESS else None, None
```
Then in the calling code:
```
ciphertext, shared_secret = encap_secret(...)
```
The encap_secret() docs just need a tweak to explain what is returned.

Reply:
Good idea; I didn't know it was possible to return two values...

Reply:
What Python does is return 1 tuple, then unpacks it into multiple vars automatically.